ch_pipeline/core/dataquery.py

```diff
@@ -655,6 +658,111 @@ def next(self, ts):
+
+            # Check to see if this timestream has a dataset id, and assign it
+            # if it does. If there is no dataset_id, we will just assume a
+            # default mapping
+            if (did_flag := ts.flags.get("dataset_id", None)) is not None:
+                # Cast bytearry type to unicode if needed
+                did_flag = memh5.ensure_unicode(did_flag[:])
+                # Get only unique dataset ids to reduce number of lookups. Ignore
+                # the null dataset at the 0th index
+                unique_dataset_ids = np.unique(did_flag)[1:]
+
+                if len(unique_dataset_ids) != 0:
+                    fmaps = {}
+                    # Cast each entry to a python string due to a bug with mySQL
+                    # typ conversion
+                    unique_dataset_ids = [str(dsid) for dsid in unique_dataset_ids]
+
+                    # Get the frequency map for each unique dataset_id. They should all
+                    # have the same frequency map - otherwise, raise an error
+                    for dsid in unique_dataset_ids:
+                        d = ds.Dataset.from_id(dsid)
+                        state = d.closest_ancestor_of_type(
+                            "f_engine_frequency_map"
+                        ).state
+                        fmaps[state.id] = state.data
+
+                    if len(fmaps.keys()) > 1:
+                        raise ValueError(
+                            "Multiple freqeuncy maps found for this dataset."
+                        )
+
+                    # Extract the stream_id and frequency map
+                    fmap_dict = list(fmaps.values())[0]["fmap"]
+
+                    stream_id = np.fromiter(fmap_dict.keys(), dtype=np.int64)
+                    fmap = np.array(list(fmap_dict.values()), dtype=np.int64)
+                    stream = tools.order_frequency_map_stream(fmap, stream_id)
+
+        # Broadcast to all ranks
+        stream_id = mpiutil.world.bcast(stream_id, root=0)
+        stream = mpiutil.world.bcast(stream, root=0)
+
+        # Check to see if we found a frequency map. If not, get the default mapping
+        if stream is None or stream_id is None:
+            stream, stream_id = tools.get_default_frequency_map_stream()
+            self.log.info("No frequency map found. Using default.")
+
+        # Set the frequency array for all 1024 frequencies, starting at 800 MHz
+        # The frequency map is only properly defined when all frequencies are used
+        freq = np.linspace(800, 400, 1024, endpoint=False)
+        cont = FrequencyMapSingle(copy_from=ts, freq=freq)
+
+        cont.stream[:] = stream
+        cont.stream_id[:] = stream_id
+
+        # Save into the cache for the next iteration
+        if self.cache:
+            self.log.debug("Caching input.")
+            self._cached_inputs = cont
+
+        # Make sure all nodes have container before return
+        mpiutil.world.Barrier()
```

Comment:
I don't think this final barrier is actually necessary. They should be pretty synchronized by the broadcasts above.